text/3113-bitfields.md

```diff
@@ -0,0 +1,534 @@
+If the value to be stored does not fit in the bit-field, the overflow check
+fails and the operation panics:
+
+```rust
+fn f(x: &mut X) {
+    x.a = 6; // Overflow check fails.
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The struct [grammar][struct-grammar] is modified:
+
+[struct-grammar]: https://doc.rust-lang.org/reference/items/structs.html
+
+```diff
+    | TupleStruct
+ 
+ StructStruct :
+-   struct IDENTIFIER  GenericParams? WhereClause? ( { StructFields? } | ; )
++   struct IDENTIFIER  GenericParams? WhereClause? ( { StructBody? } | ; )
+ 
+ TupleStruct :
+    struct IDENTIFIER  GenericParams? ( TupleFields? ) WhereClause? ;
+ 
+-StructFields :
+-   StructField (, StructField)* ,?
++StructBody :
++   StructBodyElement (, StructBodyElement)* ,?
++
++StructBodyElement :
++     StructField
++   | UnnamedStructElement
+ 
+ StructField :
+    OuterAttribute*
+    Visibility?
+    IDENTIFIER : Type
+ 
++UnnamedStructElement :
++   OuterAttribute*
++   Visibility?
++   `_` : Type
++
+ TupleFields :
+    TupleField (, TupleField)* ,?
+```
+
+UnnamedStructElements look syntactically like StructFields except that their
+names must be `_` which cannot be the name of a StructField. Semantically, they
+are not fields of the struct they are contained in.
+
+UnnamedStructElements are used when determining the layout of the struct but are
+otherwise ignored. Since they are not fields, they cannot be accessed, do not
+appear in the construction of a struct, etc.
+
+In the output of rustdoc, UnnamedStructElements are not distinguished from
+StructFields except that their name is `_`.
+
+The attribute `bits(N)` is added.  This attribute can only be applied to
+StructBodyElements of `repr(C)` structs. Such a StructBodyElement is
+called a *bit-field*. `N` must be an integer literal. `N` is called the *width*
+of the bit-field. A bit-field must have one of the types
+
+- `bool`,
+- `u8`, `u16`, `u32`, `u64`, `u128`, `usize`,
+- `i8`, `i16`, `i32`, `i64`, `i128`, `isize`
+
+modulo type aliases.
+
+`N` must be in the interval `[0, 8 * size_of::<T>()]` where `T` is the
+type of the bit-field.
+
+The attribute `bitfield(0)` can only be applied to UnnamedStructElements. An
+UnnamedStructElement must be a bit-field.
+
+Note that, despite being called a bit-*field*, a bit-field is not necessarily a
+field. To disambiguate, when talking about bit-field fields, we explicitly call
+them "bit-field fields".
+
+Each field annotated with `bits(N)` occupies `N` bits of storage.
+
+When reading and writing a bit-field field with type `bool`, `bool` is treated
+like `uM` where `M = 8 * size_of::<bool>()` with `true` corresponding to `1` and
+`false` corresponding to `0`.
+
+When a field annotated with `bits(N)` is read, the value has the type
+of the field and the behavior is as follows:
+
+- The `N` bits of storage occupied by the bit-field are read.
```

Comment:
I don't think this is correct as written. If N is 7 as then we must read (at least 8 bits). 

I think it would be better to speak about what won't be read when reading from a bit field.